Sources/PackageGraph/DependencyResolver.swift

```diff
@@ -283,20 +321,38 @@ struct PackageContainerConstraintSet<C: PackageContainer>: Collection {
         versionRequirement: VersionSetSpecifier, for identifier: Identifier
     ) -> PackageContainerConstraintSet<C>?
     {
-        let intersection = self[identifier].intersection(versionRequirement)
-        if intersection == .empty {
-            return nil
+        switch self[identifier] {
+        case .versionSet(let versionSet):
+            let intersection = versionSet.intersection(versionRequirement)
+            if intersection == .empty {
+                return nil
+            }
+            var result = self
+            result.constraints[identifier] = .versionSet(intersection)
+            return result
+        case .unversioned:
+            return self
+        }
+    }
+    private func merging(
```

Comment:
I wonder if it makes sense for this to be separate functions, since the implementations are so related?